src/main/java/edu/umich/its/cpm/Utils.java

```diff
@@ -443,19 +443,46 @@ public static String getUserLoginId(HttpServletRequest request, Environment env)
 		boolean isTestUrlEnabled = Boolean.valueOf(env.getProperty(ALLOW_USER_URLOVERRIDE));
 		String testUserInSession = (String) request.getSession().getAttribute(TEST_USER);
 
-		if (isTestUrlEnabled && testUser != null) {
+		//become user case
+		if (isTestUrlEnabled && testUser != null && remoteUser != null && testUser != remoteUser) {
+			if (isCurrentUserCPMAdmin(remoteUser, env)) {
+				request.getSession().setAttribute(TEST_USER, testUser);
+				user = testUser;
+				log.debug("BECOME USER " + user);
+			} else {
+				user = remoteUser;
+				log.error("the user " + user + " is unauthorized to become user");
+			}
+			return user;
+		}
+		//become user case
+		if (isTestUrlEnabled && testUserInSession != null && remoteUser != null && testUserInSession != remoteUser) {
+			if (isCurrentUserCPMAdmin(remoteUser, env)) {
+				user = testUserInSession;
+				log.debug("BECOME USER IN SESSION: " + user);
+			} else {
+				user = remoteUser;
+				log.error("the user " + user + " is unauthorized to become user");
+			}
+
+			return user;
+		}
+        // when no cosign enabled
```

Comment:
This isn't about cosign being enabled, it's about whether or not the container takes responsibility to manage 
authentication and setting the right value for REMOTE_USER.  It could easily be shib (or some other approach) that does that check and fills in REMOTE_USER.  This is important to be clear about since someone may think that there is something special about Cosign and there isn't.